fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -696,4 +712,38 @@ private Timestamp formatInstant(final Instant instant) {
         timestamp.setNanos(0);
         return timestamp;
     }
+
+    /**
+     * Private class to back a stream with a paged DB query.
+     *
+     * If this needs to be run in parallel we will have to override trySplit() and determine a good method to split on.
+     */
+    private class ContainmentIterator extends Spliterators.AbstractSpliterator<String> {
+        final Queue<String> children = new LinkedBlockingQueue<>();
+        final AtomicInteger numOffsets = new AtomicInteger(0);
+        final String queryToUse;
+        final MapSqlParameterSource parameterSource;
+
+        public ContainmentIterator(final String query, final MapSqlParameterSource parameters) {
+            super(Long.MAX_VALUE, Spliterator.ORDERED);
+            queryToUse = query;
+            parameterSource = parameters;
+        }
+
+        @Override
+        public boolean tryAdvance(final Consumer<? super String> action) {
+            try {
+                action.accept(children.remove());
+            } catch (final NoSuchElementException e) {
+                parameterSource.addValue("offSet", numOffsets.getAndIncrement() * containsLimit);
+                children.addAll(jdbcTemplate.queryForList(queryToUse, parameterSource, String.class));
```

Comment:
Out of curiosity, does the ordering ever matter? I ask because this implementation is trying to be thread-safe, but I'm not sure if it is entirely. Is it possible that `tryAdvance` could be called concurrently? If so, then this code will have a race to load the next page. Assuming there are sufficient items remaining, the competing threads will correctly load different pages, but the order they are added to the queue will be non-deterministic.

All that to say, I'm not sure if this implementation needs to be thread-safe. I do not believe that we have any code that uses parallel streaming features.

Reply:
My understanding, which is new, is that to allow your stream to be operated on in parallel you override the `trySplit` method which handles splitting the stream up into chunks for individual threads. 

In this case, no we don't operate on it in parallel ever. At least not right now so order is not really important except in that we need the query to generate the same order so our OFFSET and LIMIT work properly.

Reply:
In that case, do we need to use `AtomicInteger`?

Reply:
Ohhhhhhhhhh. No, I needed that when I had this class just embedded in the `StreamSupport.generate`. I can change that to an `int`